### PR TITLE
Pulls/1.0/refactor

### LINKED ISSUE #31: List modules
As a CMS user
I want to know what modules are part of my site build 
So that I can understand how the feature/ functionality sets of the site are provided

## Acceptance criteria

- Module is listed in plain English (for example "Translatable" instead of silverstripe/translatable)
- Module listing has a short description of the functionality
- Meets the SilverStripe commercially supported module standard
- Has unit tests (see above)
- The existing functionality in this module is converted to a "Report", rather than a standalone ModelAdmin
- The list only shows SilverStripe modules
- The module doesn't **rely on** filesystem commands, but uses Composer packages to gather required data as much as possible

## Tasks

- [x] Write this functionality for SilverStripe 3 (see upgrade to 4 at #42)
- [x] Convert from a ModelAdmin into a Report
- [ ] Bring up to the SilverStripe commercially supported module standard (incl. unit tests, code style, documentation)
- [x] Implement the Composer PHP SDK and Packagist API as required to use this functionality

Designs: https://projects.invisionapp.com/share/ASGU6A7RT8U#/screens/291050684

### LINKED ISSUE #37: Ensure report is available as a CSV export
As a Product Owner
I want to be able to generate a report from the Site Summariser
So that I can do further analysis and share the information with a wider group of people

## Acceptance criteria

- .csv export with complete information table is available
- Report data is in human discernable format

### LINKED ISSUE #46: Migrate ComposerPackageVersion from silverstripe-composer-versions module
The ComposerPackageVersion DataObject is "a DataObject to allow saving the information in the db to use the information in various ways".

We should move this DataObject into the silverstripe-maintenance module. The silverstripe-composer-versions module purely provides this DataObject, so could comfortably be merged into this main repository.

This would mean that the other modules, e.g. silverstripe-composer-update-checker, would be able to provide additional value via a DataExtension on this object (see https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/25)

### PR DESCRIPTION
**Pull request by NightJar**
Refactor a lot of the functionality and combine some base functionality previously provided by https://github.com/bringyourownideas/silverstripe-composer-versions as this was a bit of a stretch in terms of separation, as all other functionality provided by sister modules to it essentially rely on it.

~~This work was based on #44, and therefore includes it.~~ _Rebased_

resolves #31 
resolves #46 
resolves #37

### CODE REVIEW

File: src/Reports/SiteSummary.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+class SiteSummary extends SS_Report
```
Comment (robbieaverill) [inline src/Reports/SiteSummary.php]: Can you please add a class doc block? Also try and use namespaces wherever you can, I think that's largely anything that isn't a DataObject at the moment

File: src/Reports/SiteSummary.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        return _t(__CLASS__ . '.TITLE', 'Installed modules');
+    }
+
+    public function description()
+    {
+        return _t(
+            __CLASS__ . '.DESCRIPTION',
+            <<<DESC
+            Provides information about what SilverStripe modules are installed,
```
Comment (robbieaverill) [inline src/Reports/SiteSummary.php]: heredoc will include the whitespace here, please bring it all the way back to the left

**Comment by robbieaverill** [comment]
We'll need to add the reports module as a dependency too

File: src/Model/Package.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * This is a DataObject to cache the a package's information for later use.
+ */
+class Package extends DataObject
+{
+    private static $db = [
+        'Name' => 'Varchar(255)',
+        'Description' => 'Varchar(255)',
+        'Version' => 'Varchar(255)',
+        // 'Direct' => 'Boolean',
+    ];
```
Comment (robbieaverill) [inline src/Model/Package.php]: Is this the replacement for ComposerPackageVersion from the silverstripe-composer-versions module? E.g. does this change fix https://github.com/bringyourownideas/silverstripe-maintenance/issues/46 ?

File: src/Reports/SiteSummary.php (hunk 2 above)
Comment (NightJar) [reply src/Reports/SiteSummary.php]: Whitespace should collapse in HTML output (was my logic to leaving it in), but it's not unreasonable to remove it :)

**Comment by NightJar** [comment]
Hmm, don't merge yet, issues with namespaced reports :<

File: src/Reports/SiteSummary.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public function description()
+    {
+        return _t(
+            __CLASS__ . '.DESCRIPTION',
+            <<<DESC
+Provides information about what SilverStripe modules are installed,
+giving an insight to project statistics such as how big the installation is,
+what it would take to upgrade it, and what funcitonality is available
```
Comment (robbieaverill) [inline src/Reports/SiteSummary.php]: Typo in functionality

File: tests/Reports/Package.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    Name: pretend/uptodate
+    Description: An up to date fixture for testing
+    Version: 1.0.0
+    Latest: 1.0.0
+    Latest-status: up-to-date
+    Vendor: pretened
+    Package: uptodate
+    NiceName: uptodate
+    Upgradable: 0
+    Direct: 1
```
Comment (robbieaverill) [inline tests/Reports/Package.yml]: There's a bunch of keys here that don't exist in Package - can they be removed?

File: tests/Tasks/composer.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    
+}
```
[the remark this answers is not in the document]
Comment (NightJar) [reply tests/Tasks/composer.json]: They are test fixtures for `ComposerLoader`, in order to load fake dependencies for testing. `ComposerLoader` requires both `.json` and `.lock` to be present though, so this empty file is provided as a test double; stub.

File: tests/Tasks/UpdatePackageInfoTest.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+use SapphireTest;
+use BringYourOwnIdeas\Maintenance\Tasks\UpdatePackageInfo;
+
+class UpdatePackageInfoTest extends SapphireTest
+{
+    public function testGetPackageInfo()
+    {
+        $loader = new ComposerLoader(__DIR__);
+        $processor = new UpdatePackageInfo;
+        $output = $processor->getPackageInfo($loader->getLock()->packages);
```
[the remark this answers is not in the document]
Comment (NightJar) [reply tests/Tasks/UpdatePackageInfoTest.php]: I chose this approach as to avoid a swing of service lookups, resulting in more of an integration test than a unit test. In this manner if someone were to wish to replace the composer loader with some kind of test stub, that would be easily accomplished - `ComposerLoader` is not the unit under test here.

File: tests/Tasks/composer.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply tests/Tasks/composer.json]: You aren't testing the ComposerLoader here though, maybe you need to mock it? You can disable the constructor which is the only place where it performs filesystem calls

File: src/Tasks/UpdatePackageInfo.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            // Convert object to array, with Capitalised keys
+            $package = get_object_vars($package);
+            return array_combine(
+                array_map('ucfirst', array_keys($package)),
+                $package
+            );
+        };
+
+        $packageList = array_map($formatInfo, $packageList);
+        $this->extend(__FUNCTION__, $packageList);
```
[the remark this answers is not in the document]
Comment (NightJar) [reply src/Tasks/UpdatePackageInfo.php]: I think core is inconsistent in this manner. Some things follow that naming pattern, some do not. I'm not fussed either way, would you prefer it to be `update`?

File: tests/Tasks/UpdatePackageInfoTest.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply tests/Tasks/UpdatePackageInfoTest.php]: >ComposerLoader is not the unit under test here.
Exactly - it should be there as a mocked dependency of UpdatePackageInfo if it needs to be (`$this->getMock(ComposerLoader::class)`)

File: src/Tasks/UpdatePackageInfo.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                ' and information associated with each.'
+        );
+    }
+
+    /**
+     * Update database cached information about this site.
+     */
+    public function run($request)
+    {
+        $loader = new ComposerLoader;
```
[the remark this answers is not in the document]
Comment (NightJar) [reply src/Tasks/UpdatePackageInfo.php]: I think I can meet you half way. I'll update, see what you think then :)

**Comment by NightJar** [comment]
Screenshot of result:

image

@clarkepaul :)

**Comment by clarkepaul** [comment]
Where do we get the snippet of text which explains each module @NightJar, doesn't seem to be coming from docs. 

Do we force sentence case? as CMS is coming out as Cms, I take it we don't have any proper module name?

**Comment by NightJar** [comment]
The information comes from what is available (via composer). I am forcing "Name Case", but otherwise there's not too much to be done about the info supplied (i.e. the description). Only formatting. @clarkepaul

**Comment by clarkepaul** [comment]
So if we don't force "Name case" is it all lower case?

File: src/Tasks/UpdatePackageInfo.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Tasks/UpdatePackageInfo.php]: Yes

**Comment by robbieaverill** [comment]
>So if we don't force "Name case" is it all lower case? @NightJar

Yeah - the "plain English name" is only available as a computed value of the package name in composer.json, which is things like silverstripe/framework, silverstripe/translatable etc. The logic currently is uppercasing the latter half e.g. Framework and Transtable. Not much we can do with "Cms", other than add an extra meta tag to composer.json to contain this information in a structured way, or we could maybe keep the whole thing lowercased =)

File: src/Model/Package.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Describes an installed composer package version.
+ */
+class Package extends DataObject
+{
+    private static $db = [
+        'Name' => 'Varchar(255)',
+        'Description' => 'Varchar(255)',
+        'Version' => 'Varchar(255)',
+        "Type" => 'Varchar(255)',
```
Comment (robbieaverill) [inline src/Model/Package.php]: Any reason for the mix of quote types?

File: src/Model/Package.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    ];
+
+    private static $summary_fields = [
+        'Title',
+        'Description',
+        'Version',
+    ];
+
+    public function getTitle()
```
Comment (robbieaverill) [inline src/Model/Package.php]: PHPDoc please

File: src/Model/Package.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        $niceName = preg_replace('#^[^/]+/(silverstripe-)?#', '', $this->Name);
+        $niceName = explode('-', $niceName);
+        $niceName = implode(' ', $niceName);
+        $niceName = ucwords($niceName);
+        return $niceName;
+    }
+
+    public function getSummary()
+    {
+        return $this->renderWith('Package_summary');
```
Comment (robbieaverill) [inline src/Model/Package.php]: Out of interest, assuming this has HTML in it, how does this look when exporting to CSV?

File: src/Tasks/UpdatePackageInfo.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+ */
+class UpdatePackageInfo extends BuildTask
+{
+
+    private $composerLoader;
+
+    public function __construct($composerLoader = null)
+    {
+        parent::__construct();
+        $this->composerLoader = $composerLoader ?: Injector::inst()->create(ComposerLoader::class);
```
Comment (robbieaverill) [inline src/Tasks/UpdatePackageInfo.php]: Please use Injector dependency injection (as previously noted)

```php
private static $dependencies = ['ComposerLoader' => '%$BringYourOwnIdeas\\Maintenance\\Util\\ComposerLoader'];
```

File: src/Tasks/UpdatePackageInfo.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        return _t(
+            __CLASS__ . '.DESCRIPTION',
+            'Repopulates installation summary, listing installed modules'.
+                ' and information associated with each.'
+        );
+    }
+
+    /**
+     * Update database cached information about this site.
+     */
```
Comment (robbieaverill) [inline src/Tasks/UpdatePackageInfo.php]: Missing PHPdoc param

File: tests/Tasks/UpdatePackageInfoTest.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+use BringYourOwnIdeas\Maintenance\Util\ComposerLoader;
+use SapphireTest;
+use BringYourOwnIdeas\Maintenance\Tasks\UpdatePackageInfo;
+
+class UpdatePackageInfoTest extends SapphireTest
+{
+    public function testGetPackageInfo()
+    {
+        $loader = $this->createMock(ComposerLoader::class);
```
Comment (robbieaverill) [inline tests/Tasks/UpdatePackageInfoTest.php]: createMock is available in PHPUnit 5 but SilverStripe 3 uses 3.7 by default. Can you please switch to `$this->getMockBuilder(ComposerLoader::class)->setMethods(['getLock'])->getMock()` which is B/C

File: tests/Tasks/UpdatePackageInfoTest.php [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        $loader->method('getLock')->willReturn(json_decode(<<<LOCK
```
Comment (robbieaverill) [inline tests/Tasks/UpdatePackageInfoTest.php]: You may as well just define this as a PHP array and skip the decoding

File: tests/Tasks/UpdatePackageInfoTest.php [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+            "description": "A faux package from a mocked composer.lock for testing purposes",
+            "version": "1.0.0"
+        }
+    ]
+}
+LOCK
+        ));
+        $processor = new UpdatePackageInfo;
+        $output = $processor->getPackageInfo($loader->getLock()->packages);
+        $this->assertTrue(is_array($output));
```
Comment (robbieaverill) [inline tests/Tasks/UpdatePackageInfoTest.php]: You can use `$this->assertInternalType('array', $output)` (minor comment)

File: tests/Tasks/UpdatePackageInfoTest.php [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        $this->assertCount(1, $output);
+        $this->assertArrayHasKey('Name', $output[0]);
+        $this->assertArrayHasKey('Description', $output[0]);
+        $this->assertArrayHasKey('Version', $output[0]);
+        $this->assertEquals('fake/package', $output[0]['Name']);
```
Comment (robbieaverill) [inline tests/Tasks/UpdatePackageInfoTest.php]: Prefer `$this->assertSame()` where possible (minor comment)

File: tests/Tasks/UpdatePackageInfoTest.php [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    }
```
Comment (robbieaverill) [inline tests/Tasks/UpdatePackageInfoTest.php]: Maybe you could replace all of these assertions with something like this:

```php
$this->assertSame([
    'Name' => 'fake/package',
    'Description' => 'A faux package from a mocked composer.lock for testing purposes',
    'Version' => '1.0.0',
], $output[0]);
```

File: src/Model/Package.php (hunk 10 above)
Comment (NightJar) [reply src/Model/Package.php]: None, a side effect of me copying existing values for new keys I think. Can fix.

File: src/Model/Package.php (hunk 12 above)
Comment (NightJar) [reply src/Model/Package.php]: It looks like a string of HTML in a cell ;)
Thus it isn't included in the default `summary_fields` :)

File: tests/Tasks/UpdatePackageInfoTest.php (hunk 16 above)
Comment (NightJar) [reply tests/Tasks/UpdatePackageInfoTest.php]: Yeah, I thought to do it that way - the major issue is that objects are actually objects (as opposed to keyed array elements). So far as I'm aware there's no shorthand for an inline object definition in PHP.

Comment (ScopeyNZ) [reply tests/Tasks/UpdatePackageInfoTest.php]: You can typecast with `object` eg: https://3v4l.org/LppQB

Comment (NightJar) [reply tests/Tasks/UpdatePackageInfoTest.php]: Yeah, I also considered that... but I then also though second of it in that it sort of results in obfuscating the mocked input. Leaving it as a readable JSON string I thought was more readable :)

Thoughts?

File: src/Tasks/UpdatePackageInfo.php [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+     * @config
+     */
+    private static $dependencies = [
+        'ComposerLoader' => '%$BringYourOwnIdeas\\Maintenance\\Util\\ComposerLoader'
+    ];
+
+    /**
+     * @var ComposerLoader
+     */
+    private $composerLoader;
```
Comment (robbieaverill) [inline src/Tasks/UpdatePackageInfo.php]: Please use protected over private

File: src/Tasks/UpdatePackageInfo.php [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /**
+     * Update database cached information about this site.
+     *
+     * @param SS_HTTPRequest $request unused, can be null (must match signature of parent function).
+     */
+    public function run($request)
+    {
+        $packages = $this->getPackageInfo($this->composerLoader->getLock()->packages);
```
Comment (robbieaverill) [inline src/Tasks/UpdatePackageInfo.php]: Please dogfood the internal API: `$this->getComposerLoader()`

File: templates/Package_summary.ss [hunk 22]
```diff
@@ -0,0 +1,4 @@
+<div class="package-summary">
+    <h5 class="package-summary__title">$Title</h5>
```
Comment (robbieaverill) [inline templates/Package_summary.ss]: We need to ensure that these variables are escaped - `$Title.XML` and `$Description.XML` - this casting is default in SS4 but needs to be enforced manually in SS3

File: tests/Reports/SiteSummaryTest.php [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+        $firstRecord = $records->first();
+        $this->assertInstanceOf(Package::class, $firstRecord);
+        $this->assertStringStartsWith('pretend/', $firstRecord->Name);
+    }
+
+    public function testOnlySilverStripeModulesAreShown()
+    {
+        $summaryReport = new SiteSummary;
+        $records = $summaryReport->sourceRecords(null);
+        $this->assertEquals(3, $records->count());
```
Comment (robbieaverill) [inline tests/Reports/SiteSummaryTest.php]: `$this->assertCount(3, $records);`

**Comment by NightJar** [comment]
Updated; @clarkepaul 

image

File: src/Model/Package.php [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+    ];
+
+    /**
+     * Strips vendor and 'silverstripe-' prefix from Name property
+     * @return string More easily digestable module name for human consumers
+     */
+    public function getTitle()
+    {
+        return preg_replace('#^[^/]+/(silverstripe-)?#', '', $this->Name);
+    }
```
Comment (robbieaverill) [inline src/Model/Package.php]: This definitely needs a unit test with a bunch of different inputs, but can be done after merge

File: templates/Package_summary.ss [hunk 25]
```diff
@@ -0,0 +1,3 @@
+<strong class="package-summary__title">$Title.XML</strong>
+<% if $Description %><br /><% end_if %>
```
Comment (robbieaverill) [inline templates/Package_summary.ss]: Can you wrap it in a span and break the line with CSS instead of putting <br> tags in please?

File: src/Model/Package.php (hunk 24 above)
Comment (NightJar) [reply src/Model/Package.php]: done.

File: tests/Model/PackageTest.php [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+class PackageTest extends SapphireTest
+{
+    public function testTitleFormatsNameCorrectly()
+    {
+        $nameTests = [
+            'pretendvendor/silverstripe-prefixedpackage' => 'prefixedpackage',
+            'pretend-vendor/silverstripe-hyphen-package' => 'hyphen-package',
+            'vendor/package' => 'package',
+            'vendor/hyphenated-package' => 'hyphenated-package',
+            'silverstripe/silverstripe-silverstripe-thing' => 'silverstripe-thing'
```
Comment (robbieaverill) [inline tests/Model/PackageTest.php]: Can you add a silverstripe-themes vendor example too? Also this is a perfect scenario to use a data provider :-)
